src/plugins/memdump/userhook.cpp

```diff
@@ -0,0 +1,718 @@
+
+            if (dll_name)
+                vmi_free_unicode_str(dll_name);
+        }
+        else
+        {
+            drakvuf_release_vmi(drakvuf);
+        }
+    }
+
+    if (hctx)
+    {
+        return perform_hooking(drakvuf, info, plugin, hctx);
+    }
+
+    plugin->destroy_trap(drakvuf, info->trap);
+    return VMI_EVENT_RESPONSE_NONE;
+}
+
+static event_response_t map_view_of_section_hook_cb(drakvuf_t drakvuf, drakvuf_trap_info_t* info)
+{
+    auto plugin = get_trap_plugin<memdump>(info);
+    if (!plugin)
+        return VMI_EVENT_RESPONSE_NONE;
+
+    auto trap = plugin->register_trap<memdump, map_view_of_section_result_t<memdump>>(
+                    drakvuf,
+                    info,
+                    plugin,
+                    map_view_of_section_ret_cb,
+                    breakpoint_by_pid_searcher());
+    if (!trap)
+        return VMI_EVENT_RESPONSE_NONE;
+
+    auto data = get_trap_params<memdump, map_view_of_section_result_t<memdump>>(trap);
+    if (!data)
+    {
+        plugin->destroy_plugin_params(plugin->detach_plugin_params(trap));
+        return VMI_EVENT_RESPONSE_NONE;
+    }
+
+    data->set_result_call_params(info, drakvuf_get_current_thread(drakvuf, info));
+
+    // IN HANDLE SectionHandle
+    data->section_handle = drakvuf_get_function_argument(drakvuf, info, 1);
+    // IN HANDLE ProcessHandle
+    data->process_handle = drakvuf_get_function_argument(drakvuf, info, 2);
+    // IN OUT PVOID *BaseAddress
+    data->base_address_ptr = drakvuf_get_function_argument(drakvuf, info, 3);
+
+    return VMI_EVENT_RESPONSE_NONE;
+}
+
+/**
+ * As we may accidentally trigger an exception in the kernel by using vmi_request_page_fault,
+ * we hook KiSystemServiceHandler to account for that situation. Inside this hook,
+ * we check if it was "our fault" and if so, we forcefully return EXCEPTION_CONTINUE_EXECUTION.
+ * In any other case, we just pass the control to the original exception handler.
+ */
+static event_response_t system_service_handler_hook_cb(drakvuf_t drakvuf, drakvuf_trap_info_t* info)
+{
+    printf("[USERHOOK] Entered system service handler\n");
+
+    auto plugin = get_trap_plugin<memdump>(info);
+    if (!plugin)
+        return VMI_EVENT_RESPONSE_NONE;
+
+    GSList* it = plugin->hooking_contexts;
+    uint32_t thread_id;
+
+    if (!drakvuf_get_current_thread_id(drakvuf, info, &thread_id))
+    {
+        printf("[USERHOOK] Failed to get thread id in system service handler!\n");
+        return VMI_EVENT_RESPONSE_NONE;
+    }
+
+    while (it)
+    {
+        hooking_ctx_t* hctx = (hooking_ctx_t*)it->data;
+
+        if (hctx->thread_id == thread_id)
+        {
+            break;
+        }
+
+        it = it->next;
+    }
+
+    if (!it)
+    {
+        printf("[MEMDUMP] Shouldnt catch\n");
+        return VMI_EVENT_RESPONSE_NONE;
+    }
+
+    printf("[MEMDUMP] jump to ret gadget\n");
+    info->regs->rip = plugin->ret_gadget;
```

Comment:
I would say it would be easier to just set the rip to address that would be popped from the stack and adjusting the stack pointer (ie. emulate the `RET`) instead of searching for this gadget and relying on it not being patched out under your feet.

Reply:
done